File: src/main/java/com/github/sshaddicts/lucrecium/runner.java
package com.github.sshaddicts.lucrecium;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.sshaddicts.lucrecium.datasets.ImageDataSet;
import com.github.sshaddicts.lucrecium.imageProcessing.ImageProcessor;
import com.github.sshaddicts.lucrecium.imageProcessing.Imshow;
import com.github.sshaddicts.lucrecium.imageProcessing.containers.CharContainer;
import com.github.sshaddicts.lucrecium.neuralNetwork.RichNeuralNet;
import com.github.sshaddicts.lucrecium.neuralNetwork.TextRecognizer;
import com.github.sshaddicts.lucrecium.util.FileInteractions;
import org.deeplearning4j.api.storage.StatsStorage;
import org.deeplearning4j.ui.api.UIServer;
import org.deeplearning4j.ui.stats.StatsListener;
import org.deeplearning4j.ui.storage.InMemoryStatsStorage;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.opencv.core.Core;
import org.opencv.core.Mat;
import org.opencv.imgcodecs.Imgcodecs;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

public class runner {

    static {
        System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
    }

    private static final List<String> filenames = FileInteractions.getFileNamesFromDir("dataSources/greatData");
    private static final Logger log = LoggerFactory.getLogger(runner.class);
    private static List<String> labels;

    public static void main(String[] args) throws IOException {
        ObjectMapper mapper = new ObjectMapper();


        recognizeDigits();

    }

    public static void recognizeDigits() throws IOException {
        ImageProcessor processor = new ImageProcessor("testCase/numbers31.jpg");


        TextRecognizer recognizer = new TextRecognizer("netFile");

        List<CharContainer> textRegions = processor.findTextRegions(ImageProcessor.NO_MERGE);

        String text = recognizer.getText(textRegions);

        System.out.println(text);
        Imshow.show(processor.getImage(), "processed");

        Imgcodecs.imwrite("overlay.png", processor.getOverlay());
    }

    public static void testNetwork() throws IOException {
        RichNeuralNet net = new RichNeuralNet(10);
        net.loadNetwork("netFile");

        ImageDataSet dataSet = new ImageDataSet(3, 1);
        dataSet.initFromDirectory("testData/TrainF");

        DataSetIterator iterator = dataSet.getIterator();

        while (iterator.hasNext()) {
            int[] output = net.getNet().predict(iterator.next().getFeatureMatrix());
            System.out.println(Arrays.toString(output));
        }
        System.out.println(2 + 2);
    }

    public static void trainNetwork(int epochs) throws IOException {
        ImageDataSet dataSet = new ImageDataSet(10, 28);
        dataSet.initFromDirectory("testData/testF");

        RichNeuralNet net = new RichNeuralNet(10);

        net.initLenetMnist();

        UIServer uiServer = UIServer.getInstance();
        StatsStorage statsStorage = new InMemoryStatsStorage();
        uiServer.attach(statsStorage);
        net.getNet().setListeners(new StatsListener(statsStorage));

        for (int i = 0; i < epochs; i++) {
            net.train(dataSet.getIterator());
        }

        net.saveNetwork("netFile");

        labels = dataSet.getRecordReader().getLabels();
    }

}

File: src/main/java/com/github/sshaddicts/lucrecium/Runner.kt
package com.github.sshaddicts.lucrecium

import com.github.sshaddicts.lucrecium.datasets.ImageDataSet
import com.github.sshaddicts.lucrecium.imageProcessing.ImageProcessor
import com.github.sshaddicts.lucrecium.imageProcessing.Imshow
import com.github.sshaddicts.lucrecium.neuralNetwork.RichNeuralNet
import com.github.sshaddicts.lucrecium.neuralNetwork.TextRecognizer
import com.github.sshaddicts.lucrecium.util.FileInteractions
import org.apache.commons.io.FileUtils
import org.opencv.core.Core
import org.opencv.imgcodecs.Imgcodecs
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.util.*

object Runner {

    //private val filenames = FileInteractions.getFileNamesFromDir("dataSources/newData")
    //private val greatFile = FileInteractions.getFileNamesFromDir("dataSources/greatData")
    //private val netDir = "outputData/productionDataSet2"
    private val log = LoggerFactory.getLogger(this.javaClass)

    val labels: List<String> = listOf(
            "0","1","2","3","4","5",
            "6","7","8","9","dot","slash",
            "Є","І","Ї","А","Б","В",
            "Г","Д","Е","Ж","З","И",
            "Й","К","Л","М","Н","О",
            "П","Р","С","Т","У","Ф",
            "Х","Ц","Ч","Ш","Щ","Ь",
            "Ю","Я",
            "а","б","в","г","д","е",
            "ж","з","и","й","к","л",
            "м","н","о","п","р","с",
            "т","у","ф","х","ц","ч",
            "ш","щ","ь","э","ю","я",
            "є","і","ї","ґ"
    )

    init {
        System.loadLibrary(Core.NATIVE_LIBRARY_NAME)
    }

    @Throws(IOException::class, NoSuchMethodException::class, InvocationTargetException::class, IllegalAccessException::class)
    @JvmStatic
    fun main(args: Array<String>) {

        val numb = 3
        //trainNetwork(1, outputLabelCount);

        testNetworkOnImage(numb, labels)
        println("done")
    }

    @Throws(IOException::class)
    fun recognizeDigits() {
        val image = ImageProcessor.loadImage("testCase/numbers.jpg")
        val processor = ImageProcessor()

        val recognizer = TextRecognizer("netFileModLetters.lucrecium")

        val (chars) = processor.findTextRegions(image, ImageProcessor.NO_MERGE, false)

        val text = recognizer.recognize(chars, ArrayList())

        println(text)
    }

    @Throws(IOException::class)
    fun tryCatch() {
        val bytes = FileUtils.readFileToByteArray(File("testCase/numbers31.jpg"))

        val image = ImageProcessor.loadImage(bytes)

        Imshow.show(image, "1")

        val processor = ImageProcessor()

        val (chars, overlay) = processor.findTextRegions(image)

        val recognizer = TextRecognizer("netFile")

        val recognize = recognizer.recognize(chars, ArrayList())

        println(recognize)
        Imshow.show(chars[0].mat)
        Imshow.show(ImageProcessor.loadImage(overlay, Imgcodecs.CV_LOAD_IMAGE_UNCHANGED))
    }

    @Throws(IOException::class, NoSuchMethodException::class, InvocationTargetException::class, IllegalAccessException::class)
    fun testNetworkOnImage(num: Int, labels: List<String>) {
        val image = ImageProcessor.loadImage("test.jpg")
        val processor = ImageProcessor()

        val (chars, overlay) = processor.findTextRegions(image, ImageProcessor.NO_MERGE, true)
        println("found " + chars.size + " elements")

        val recognizer = TextRecognizer("netFileModLetters.lucrecium")
        val recognize = recognizer.recognize(chars, labels)

        for (node in recognize) {
            println(node)
        }

    }


}
